### PR TITLE
Remove inverse property from `Operation`

### PR DESCRIPTION
**Pull request by albi3ro**
Follow up to #3618  that removed the ability to set the `inverse` property.

This PR removes `Operation.inverse` completely from `pennylane`. This will probably break plugins, which we will need to fix before the release.

### CODE REVIEW

File: pennylane/operation.py [hunk 1]
```diff
@@ -1623,43 +1618,12 @@ def parameter_frequencies(self):
             "and parameter frequencies can not be computed as no generator is defined."
         )
 
-    @property
-    def inverse(self):
-        """Boolean determining if the inverse of the operation was requested."""
-        return False
-
-    def matrix(self, wire_order=None):
```
Comment (rmoyard) [inline pennylane/operation.py:1631]: Why do you remove the matrix method completely? Is not expected to stay in the abc class?

**Review by rmoyard — COMMENTED** [review]
Great to see that we get rid of inv :+1: A question and I think they are still some functions change adjoint in metric tensor:

`def _apply_operations(state, op, device, invert=False):`

File: pennylane/operation.py (hunk 1 above)
Comment (AlbertMitjans) [reply pennylane/operation.py:1631]: We have an `Operator.matrix` method that does the same thing

**Review by AlbertMitjans — COMMENTED** [review]
Amazing to finally get rid of this! 🤩 💯 

Below you can find a list of files where `inverse` or `inv` is still being used/mentioned:
- **equal.py**: Line 209.
- **is_commuting.py**: Lines 125, 126 and 129.
- **adjoint.py**: Line 338.
- **controlled.py**: Line 560.
- **exp.py**: Line 227.
- **qscript.py**: Lines 56 and 58.
- **cancel_inverses.py**: Do we need this file at all?
- **_device.py**: Lines 903-909, 934, 970-979.
- **_qubit_device.py**: Lines 679-692.

File: pennylane/ops/functions/equal.py [hunk 2]
```diff
@@ -206,7 +206,7 @@ def _equal_operators(
             if qml.math.get_interface(params_1) != qml.math.get_interface(params_2):
                 return False
 
-    return getattr(op1, "inverse", False) == getattr(op2, "inverse", False)
+    return True
```
Comment (rmoyard) [inline pennylane/ops/functions/equal.py:209]: What cases does this line now cover?

File: pennylane/transforms/qmc.py [hunk 3]
```diff
@@ -304,25 +304,31 @@ def qmc():
...
-         'depth': 30610,
-         'num_device_wires': 12,
-         'device_name': 'default.qubit.autograd',
-         'diff_method': 'backprop'}
+        'gate_types': defaultdict(int,
+                    {'RY': 7747,
+                    'CNOT': 7874,
+                    'Hadamard': 258,
+                    'CZ': 126,
+                    'Adjoint(CNOT)': 7812,
```
Comment (rmoyard) [inline pennylane/transforms/qmc.py:314]: Adjoint CNot should be CNOT, is it not simplified here?

**Review by rmoyard — COMMENTED** [review]
Two questions, then happy to approve :+1:

File: pennylane/transforms/qmc.py (hunk 3 above)
Comment (albi3ro) [reply pennylane/transforms/qmc.py:314]: `Adjoint(CNOT)` is technically supported by the default qubit, so it doesn't get decomposed. We could technically change Line 101 to be `adjoint(fn, lazy=False)`.  Or we could consider making the default behaviour "non-lazy". But I think both solutions are outside the scope of this PR.

File: pennylane/ops/functions/equal.py (hunk 2 above)
Comment (albi3ro) [reply pennylane/ops/functions/equal.py:209]: If none of the other cases have return False, then we know they are equal.

We could potentially rearrange the order of checks to end with something like:
```
return op1.hyperparameters == op2.hyperparameters
```
but I don't think it's that important of a change.

**Review by rmoyard — APPROVED** [review]
It looks good to me thanks :100:

**Pull request merged**